### PR TITLE
[WIP,RFT,RFC] generic: qca8k: add experimental patch for eth mdio

### PR DESCRIPTION
**Pull request by Ansuel**
Feature:
--
- Add support for mdio read/write in Ethernet packet.
- Add support for phy read/write in Ethernet packet.
- Add support for MIB counter in Ethernet packet.
- Bulk read/write. Read and write fdb entry in one go.

I need some stability testing and some perf stats.

DEPENDS
--
- Dsa conversion patch for ipq806x #4036 
- kernel 5.15 support patch #4748 

Current problem:
--
~This slow down init. I still need to find the correct way to understand when the tagger/dsa starts to receive/send packet. Because of this every request has to timeout and use the legacy mdio way.
When the system starts to handle packet from the switch port the mdio read/write will use the Ethernet way and will be fasted.~ (With the help of the kernel dev we develop a way to track when the master port (and the tagger) can correctly receive and handle packet)

Reason:
--
To support multicpu port, we probably need to enable assisted learning for cpu. This cause an increase in mdio read/write and this new way should speedup everything and remove the problem of assisted learning.

I would really appreciate any test with this.

Additions:
--
This also include another pr that permit to skip some redundant mdio write to reduce even more mdio use.
I should also test if we can use the same logic for the mdio read. In theory yes...

WARNING
--
This is not compile tested on a normal openwrt... I compiled this on linux kernel and I directly modified the source in the build dir. So expect some problem with patch apply.

I got some report that this apply directly to 5.15 pr.

@robimarko wonder if you can give some comments on the implementation if you have any idea on how this can be improved... 

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

### CODE REVIEW

File: target/linux/generic/pending-5.10/999-5-net-dsa-tag_qca-Add-support-for-handling-mdio-read-w.patch [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
++
++	cpu_dp = dev->dsa_ptr;
++	header = cpu_dp->priv;
++
++	mdio_ethhdr = (struct mdio_ethhdr *)skb_mac_header(skb);
++
++	header->data[0] = mdio_ethhdr->mdio_data;
++
++	/* Get the rest of the 12 byte of data */
++	memcpy(header->data + 1, skb->data, QCA_HDR_MDIO_DATA2_LEN);
```
Comment (Ansuel) [inline target/linux/generic/pending-5.10/999-5-net-dsa-tag_qca-Add-support-for-handling-mdio-read-w.patch:47]: I can check the cmd and skip this memcpy and the other set with write operation.

File: target/linux/generic/pending-5.10/999-6-net-dsa-qca8k-cache-lo-and-hi-for-mdio-write.patch [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+diff --git a/drivers/net/dsa/qca8k.c b/drivers/net/dsa/qca8k.c
+index d2c6139be9ac..64643f1e2f16 100644
+--- a/drivers/net/dsa/qca8k.c
++++ b/drivers/net/dsa/qca8k.c
+@@ -94,6 +94,48 @@ qca8k_split_addr(u32 regaddr, u16 *r1, u16 *r2, u16 *page)
+ 	*page = regaddr & 0x3ff;
+ }
+ 
++static u16 qca8k_current_lo = 0xffff;
```
Comment (Ansuel) [inline target/linux/generic/pending-5.10/999-6-net-dsa-qca8k-cache-lo-and-hi-for-mdio-write.patch:23]: We should consider a more safe value.... But considering the first thing will be checking the switch id... this should be fine...

**Comment by rbpp** [comment]
@Ansuel : Hi Ansuel. Is this PR based on dsa or swconfig? I'm willing to test it but I need to know what config to use.

**Comment by Ansuel** [comment]
it's on dsa

Il giorno ven 10 dic 2021 alle ore 22:13 rbpp ***@***.***> ha
scritto:

**Comment by rbpp** [comment]
> it's on dsa.

My question was a bit foolish, qca8k has to be on dsa... Anyway, since this PR does not contain the other required changes to work on dsa (e.g., dts changes and maybe others), do I have to apply this PR together with another one of your open PRs?

**Comment by Ansuel** [comment]
Yes sorry I rushed this and will change shortly as I'm stabilizing it with the kernel guy.
Anyway this require 5.15 and ipq806x patch 
But if you want to wait I will push a totally different series that should also works better and have more features.
Also notice that 5.15 require firewall 4.

**Comment by rbpp** [comment]
Take your time, I'll wait for your magic!

**Comment by Ansuel** [comment]
I pushed the """""""final""""""""" version of this... Hope someone can test this and give some result of stability / perf.
(this is the full conversion to eth packet
this use eth packet for
- mib counter
- mdio read/write
- phy read/write
)

File: target/linux/generic/pending-5.10/999-15-net-dsa-qca8k-cache-lo-and-hi-for-mdio-write.patch [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+diff --git a/drivers/net/dsa/qca8k.c b/drivers/net/dsa/qca8k.c
+index 375a1d34e46f..b109a74031c6 100644
+--- a/drivers/net/dsa/qca8k.c
++++ b/drivers/net/dsa/qca8k.c
+@@ -94,6 +94,48 @@ qca8k_split_addr(u32 regaddr, u16 *r1, u16 *r2, u16 *page)
+ 	*page = regaddr & 0x3ff;
+ }
+ 
++static u16 qca8k_current_lo = 0xffff;
```
Comment (e9hack) [inline target/linux/generic/pending-5.10/999-15-net-dsa-qca8k-cache-lo-and-hi-for-mdio-write.patch:24]: qca8k_current_lo/high/page must be part of private data in structure qca8k_priv. If multiple qca8k switches does exist, currently they will share this three cached values.

**Comment by rbpp** [comment]
@Ansuel : just loaded this PR on my dumb C2600 and it's working (it's with fw3, but I don't use firewall on this AP). Any test you want me to carry?

I can also test this PR on my main R7800 but I did not figure out yet how to compile a build with firewall4 in it.

**Comment by Ansuel** [comment]
You should check port stability and CPU load with traffic. What I need to
know is if the switch stability is equal.

Il Dom 12 Dic 2021, 15:07 rbpp ***@***.***> ha scritto:

**Comment by rbpp** [comment]
I upgraded my R7800 with this PR and it's operational (with fw3). Bufferbloat test was not very good though but I can't tell if it's caused by this PR or by kernel 5.15.

**Comment by Ansuel** [comment]
@rbpp do you have any stats? bufferbloat probably doesn't work as no firewall.
port dropping anything?

**Comment by rbpp** [comment]
> @rbpp do you have any stats? bufferbloat probably doesn't work as no firewall. port dropping anything?

Ports were stable, no drops in the logs. My C2600 was 3 days with this PR and everything looked good. As with my R7800, I used this PR only for a day because I did not want to have my main router exposed because of the problematic firewall, but it was otherwise stable during that period.

**Comment by Ansuel** [comment]
@rbpp was thinking, did you moved the patch to 5.15 dir or just applied this pr? Anyway the logic was wrong in the old series... I fixed it and make this more stable. Test it whatever you can :D

**Comment by rbpp** [comment]
> @rbpp was thinking, did you moved the patch to 5.15 dir or just applied this pr?

I had applied this PR + #4748 (k5.15) + #4036 (ipq806x dsa), as per your instruction. I'll test this new version in the next few days.

**Comment by KA2107** [comment]
@Ansuel Thank you for your work. Just one minor thing, "experimental" spelling is wrong in both the Pull Request title and one of the commits.

**Comment by Ansuel** [comment]
@KA2107 just notice that. Anyway I'm rushing this to make it at least compliable... I need to finish a feature and I should really start refreshing all these pr... (example the add experimental patch got merged upstream and I have to backport them to 5.15)

File: target/linux/generic/pending-5.10/999-15-net-dsa-qca8k-cache-lo-and-hi-for-mdio-write.patch (hunk 3 above)
Comment (Ansuel) [reply target/linux/generic/pending-5.10/999-15-net-dsa-qca8k-cache-lo-and-hi-for-mdio-write.patch]: Aside from the the global values that just got fixed, (still needs to be merged but i'm spending some time clearing the pr)

Caching the values looks wrong IMHO and this caching is on the regs level so we don't cache values, we cache the indirect regs used for mdio.

**Comment by rmandrad** [comment]
@Ansuel - are you using netfilter on 5.15 ? nf_log_common doesn't exist  on 5.15 anymore but can't find anything on your PR that removes the dependencies
thank you

**Comment by Ansuel** [comment]
Anyway got this merged upstream... Will close this and include backport patches in the 5.15 pr

**Comment by robimarko** [comment]
Damn, completely missed the upstream discussion lately.

Nice job.

**Comment by Ansuel** [comment]
@robimarko i'm still very tempted to propose the code split... wonder if they will accepted it anyway even if there isn't a user

**Comment by robimarko** [comment]
No way it will get accepted without a user

**Comment by Ansuel** [comment]
Anyway I feel like this should be improved... I need to understand all the build_skb logic as it looks wrong that space is allocated and free multiple time for a packet of the same size...

**Comment by robimarko** [comment]
I wouldnt worry too much, as long as its freed and not leaked its fine

**Comment by Ansuel** [comment]
it's problematic for ath79 devices as the cpu is not that fast

**Comment by robimarko** [comment]
It cant be that bad, its not like you have millions of packets per second for management.

**Comment by Ansuel** [comment]
@robimarko we have tons of packet for the state machine to check the port status 100 a seconds. but i assume they can be decreased by turning off polling.. it was enabled as the cpu port can be set to base-x but never understood why they didn't just limit it to that particular mode... But I have to investigate it... could be that polling mode can be set only on switch setup and not at the mac config call

**Pull request closed without merging**